src/components/sidenav/sidenav.js

```diff
@@ -505,5 +503,19 @@ function SidenavController($scope, $element, $attrs, $mdComponentRegistry, $q) {
   self.toggle = function() { return self.$toggleOpen( !$scope.isOpen );  };
   self.$toggleOpen = function(value) { return $q.when($scope.isOpen = value); };
 
-  self.destroy = $mdComponentRegistry.register(self, $attrs.mdComponentId);
+  // Evaluate the component id.
+  var rawId = $attrs.mdComponentId;
+  var componentId = rawId && rawId.indexOf($interpolate.startSymbol()) > -1 ?
+    $interpolate(rawId)($scope.$parent) : rawId;
+
+  // Register the component.
+  self.destroy = $mdComponentRegistry.register(self, componentId);
+
+  // Watch and update the component, if the id has changed.
+  $attrs.$observe('mdComponentId', function(id) {
+    if (id && id !== self.$$mdHandle) {
+      self.destroy();
+      self.destroy = $mdComponentRegistry.register(self, id);
```

Comment:
The `register` method returns a method that can be used to de-register the component. Here we de-register the instance, because it was using the old id, and we register it again with the new one.